### PR TITLE
target/riscv: Handle harts powering down and coming back up.

### PR DESCRIPTION
**Pull request by timsifive**

### CODE REVIEW

File: src/target/riscv/riscv-013.c [hunk 1]
```diff
@@ -587,6 +587,7 @@ static int dmi_op_timeout(struct target *target, uint32_t *data_in,
 			break;
 		} else {
 			LOG_ERROR("failed %s at 0x%x, status=%d", op_name, address, status);
+			dtmcontrol_scan(target, DTM_DTMCS_DMIRESET);
 			return ERROR_FAIL;
```
Comment (JanMatCodasip) [inline src/target/riscv/riscv-013.c]: Here you can just reset the DMI and break from the loop. The error message and function return is already handled below.

```suggestion
			dtmcontrol_scan(target, DTM_DTMCS_DMIRESET);
			break;
```

File: src/target/riscv/riscv-013.c [hunk 2]
```diff
@@ -595,6 +596,7 @@ static int dmi_op_timeout(struct target *target, uint32_t *data_in,
 
 	if (status != DMI_STATUS_SUCCESS) {
 		LOG_ERROR("Failed %s at 0x%x; status=%d", op_name, address, status);
+		dtmcontrol_scan(target, DTM_DTMCS_DMIRESET);
 		return ERROR_FAIL;
 	}
```
Comment (JanMatCodasip) [inline src/target/riscv/riscv-013.c]: No need to call `dtmcontrol_scan(target, DTM_DTMCS_DMIRESET)` here again. It has already been handled above.

The word "DMI" can be included in the error message to make it clearer to the user what type of error it actually is.

```suggestion
	if (status != DMI_STATUS_SUCCESS) {
		LOG_ERROR("Failed DMI %s at 0x%x; status=%d", op_name, address, status);
		return ERROR_FAIL;
	}
```

File: src/target/riscv/riscv-013.c [hunk 3]
```diff
@@ -619,6 +621,7 @@ static int dmi_op_timeout(struct target *target, uint32_t *data_in,
 					LOG_ERROR("Failed %s (NOP) at 0x%x; status=%d", op_name, address,
```
Comment (JanMatCodasip) [inline src/target/riscv/riscv-013.c]:
```suggestion
					LOG_ERROR("Failed DMI %s (NOP) at 0x%x; status=%d", op_name, address,
```

(There's also one more occurrence above.)

File: src/target/target.c [hunk 4]
```diff
@@ -3073,8 +3076,7 @@ static int handle_target(void *priv)
 			target_reset_examined(target);
 			retval = target_examine_one(target);
 			if (retval != ERROR_OK) {
-				LOG_TARGET_DEBUG(target, "Examination failed, GDB will be halted. "
-					"Polling again in %dms",
+				LOG_TARGET_DEBUG(target, "Examination failed. Polling again in %dms",
```
Comment (JanMatCodasip) [inline src/target/target.c:3073]: Detail: 
```suggestion
				LOG_TARGET_DEBUG(target, "Examination failed. Polling again in %d ms",
```

File: src/target/riscv/riscv-013.c (hunk 1 above)
Comment (timsifive) [reply src/target/riscv/riscv-013.c]: Done.

File: src/target/riscv/riscv-013.c (hunk 2 above)
Comment (timsifive) [reply src/target/riscv/riscv-013.c]: Done.

File: src/target/riscv/riscv-013.c (hunk 3 above)
Comment (timsifive) [reply src/target/riscv/riscv-013.c]: Done.

File: src/target/target.c (hunk 4 above)
Comment (timsifive) [reply src/target/target.c:3073]: In the OpenOCD code that I did not write, I see 3 cases of "%d ms" and 3 cases of "%dms". I'm inclined to leave this as-is, and let upstream worry about it when they get around to nitpicking whitespace in debug messages.

Comment (JanMatCodasip) [reply src/target/target.c:3073]: Agreed :)

File: src/target/riscv/riscv-013.c [hunk 5]
```diff
@@ -1741,6 +1744,18 @@ static int examine(struct target *target)
 		r->mtopi_readable = false;
 	}
 
+	riscv_reg_t dcsr;
```
Comment (GregSavin) [inline src/target/riscv/riscv-013.c]: Does this code get reached if power down/up happens after the first call to "init" command happens?

File: src/target/riscv/riscv-013.c [hunk 6]
```diff
@@ -1744,6 +1744,18 @@ static int examine(struct target *target)
...
+		return ERROR_FAIL;
+	dcsr = set_field(dcsr, CSR_DCSR_STEP, 0);
+	dcsr = set_field(dcsr, CSR_DCSR_EBREAKM, riscv_ebreakm);
+	dcsr = set_field(dcsr, CSR_DCSR_EBREAKS, riscv_ebreaks);
+	dcsr = set_field(dcsr, CSR_DCSR_EBREAKU, riscv_ebreaku);
+	dcsr = set_field(dcsr, CSR_DCSR_EBREAKVS, riscv_ebreaku);
+	dcsr = set_field(dcsr, CSR_DCSR_EBREAKVU, riscv_ebreaku);
+	if (register_write_direct(target, GDB_REGNO_DCSR, dcsr) != ERROR_OK)
+		return ERROR_FAIL;
+
```
Comment (JanMatCodasip) [inline src/target/riscv/riscv-013.c]: The same operation is performed in `riscv013_on_step_or_resume()`, so the code can be factored out to a single helper function, called multiple times.

The setting of `dcsr.ebreak*` bits should be also done in `deassert_reset`. I have recorded it as an issue: https://github.com/riscv/riscv-openocd/issues/853 .

Comment (timsifive) [reply src/target/riscv/riscv-013.c]: I've refactored. Let's leave the deassert fix for another PR.

File: src/target/riscv/riscv-013.c (hunk 5 above)
Comment (timsifive) [reply src/target/riscv/riscv-013.c]: Yes, but the mechanism is fragile. What happens is you get a failure in poll(), and then in handle_target() that failure is seen and it calls target_reset_examined().

I'll try to make this mechanism more robust.

File: src/target/riscv/riscv-013.c (hunk 6 above)
Comment (JanMatCodasip) [reply src/target/riscv/riscv-013.c]: Sure, the setting of `dcsr.ebreak* ` in `deassert_reset` is a topic for a separate PR. I've mentioned it just for completeness :) Thanks.

**Review by GregSavin — DISMISSED** [review]

**Review by JanMatCodasip — DISMISSED** [review]
Checked visually - looks all right.

**Comment by timsifive** [comment]
Force pushed to rebase, avoiding conflicts.

**Comment by timsifive** [comment]
I'd like to merge this because it makes it better. I'm working on a more robust solution (which will also help with the reset case), but it's getting complex quickly when there are multiple harts and they might halt for reasons besides that we requested them to halt.

**Review by GregSavin — APPROVED** [review]

**Pull request merged**